lib/galaxy/jobs/runners/aws.py

```diff
@@ -0,0 +1,598 @@
+""" Galaxy job runners to use Amazon AWS native compute resources, such as AWS Batch.
+"""
+import hashlib
+import json
+import logging
+import os
+import re
+import time
+from queue import Empty
+from typing import List
+
+from galaxy import model
+from galaxy.job_execution.output_collect import default_exit_code_file
+from galaxy.jobs.runners import (
+    AsynchronousJobRunner,
+    AsynchronousJobState,
+    JobState,
+)
+from galaxy.util import (
+    smart_str,
+    unicodify,
+)
+
+BOTO3_IMPORT_MSG = (
+    "The Python 'boto3' package is required to use "
+    "this feature, please install it or correct the "
+    "following error:\nImportError {msg!s}"
+)
+
+try:
+    import boto3
+
+except ImportError as e:
+    boto3 = None
+    BOTO3_IMPORT_MSG.format(msg=unicodify(e))
+
+
+__all__ = ("AWSBatchJobRunner",)
+log = logging.getLogger(__name__)
+
+STOP_SIGNAL = object()
+
+
+class AWSBatchRunnerException(Exception):
+    pass
+
+
+def to_dict(segments, v):
+    if len(segments) == 0:
+        return v
+    return {segments[0]: to_dict(segments[1:], v)}
+
+
+def _write_logfile(logfile, msg):
+    with open(logfile, "w") as fil:
+        fil.write(msg)
+
+
+def _add_galaxy_environment_variables(vcpu, memory):
+    # Set:
+    # GALAXY_SLOTS, round 0.25 vpuc to 1.
+    # GALAXY_MEMORY_MB
+    return [
+        {"name": "GALAXY_SLOTS", "value": str(int(max(vcpu, 1)))},
+        {"name": "GALAXY_MEMORY_MB", "value": str(memory)},
+    ]
+
+
+def _add_resource_requirements(destination_params):
+    rval = [
+        {"type": "VCPU", "value": str(destination_params.get("vcpu"))},
+        {"type": "MEMORY", "value": str(destination_params.get("memory"))},
+    ]
+    n_gpu = destination_params.get("gpu")
+    if n_gpu:
+        rval.append({"type": "GPU", "value": str(n_gpu)})
+    return rval
+
+
+class AWSBatchJobRunner(AsynchronousJobRunner):
+    """
+    This runner uses container only. It requires that an AWS EFS is mounted as a local drive
+    and all Galaxy job-related paths, such as objects, job_directory, tool_directory and so
+    on, are placed in the EFS drive. Via this runner, Galaxy jobs are dispatched to AWS Batch
+    for compute using the docker image specified by a Galaxy tool. As AWS EFS is designed to
+    be able to mount at multiple places with read and write capabilities, Galaxy and Batch
+    containers share the same EFS drive as a local device. Sample configurations can be found
+    in `config/job_conf.xml.sample_advanced`.
+    """
+
+    runner_name = "AWSBatchRunner"
+    RUNNER_PARAM_SPEC_KEY = "runner_param_specs"
+    JOB_NAME_PREFIX = "galaxy-"
+    # AWS Batch queries up to 100 jobs at once.
+    MAX_JOBS_PER_QUERY = 100
```

Comment:
Thank you @qiagu for working on this. Can we add this parameter to the configuration options?

Reply:
The number/limit is managed by AWS. I don't think AWS changes the number often. Also, I imagine performance would be compromised if we change it to a smaller number. Is it really necessary to allow Galaxy admin to control it via job configuration?